### PR TITLE
FM-246 Add Swagger SDK generation backend implementation

### PR DESCRIPTION
**Pull request by eunice18**

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/fhir/api/util/FHIRRESTfulGenericClient.java [hunk 1]
```diff
@@ -56,9 +56,9 @@ public static ImagingStudy readImagingStudyById(String serverBase,
 	}
 
 	public static Bundle searchWhereReferenceAndToken(String serverBase,
-	                                                  Class<DiagnosticReport> fhirResource,
-	                                                  ICriterion<ReferenceClientParam> where,
-	                                                  ICriterion<TokenClientParam> and) {
+														   Class<DiagnosticReport> fhirResource,
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/util/FHIRRESTfulGenericClient.java]: exceeds maximum length of line 120

File: api/src/main/java/org/openmrs/module/fhir/api/util/FHIRRESTfulGenericClient.java [hunk 2]
```diff
@@ -68,8 +68,8 @@ public static Bundle searchWhereReferenceAndToken(String serverBase,
 	}
 
 	public static Bundle searchWhereReference(String serverBase,
-	                                          Class<DiagnosticReport> fhirResource,
-	                                          ICriterion<ReferenceClientParam> where) {
+											  Class<DiagnosticReport> fhirResource,
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/util/FHIRRESTfulGenericClient.java]: same here

File: omod/src/main/java/org/openmrs/module/fhir/swagger/codegen/SwaggerCodeGenerator.java [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+    protected Log log = LogFactory.getLog(getClass());
+    private Map<String, String> sdkLanguages = new HashMap<>();
+    private static final String ZIP_ARCHIVE_NAME_PREFIX = "OPENMRS-FHIR-CLIENT";
+    public static final String JSON_EXTENSION = ".json";
+
+
+    public SwaggerCodeGenerator() {
+        //Supported SDK languages
+        sdkLanguages.put("java", "io.swagger.codegen.languages.JavaClientCodegen");
```
Comment (harsha89) [inline omod/src/main/java/org/openmrs/module/fhir/swagger/codegen/SwaggerCodeGenerator.java]: are we only supports these languages only? I thought according to the discussion we are supporting all available languages. Can you please check

File: omod/src/main/java/org/openmrs/module/fhir/swagger/codegen/SwaggerCodeGenerator.java [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+        sdkLanguages.put("asyncScala", "io.swagger.codegen.languages.AsyncScalaClientCodegen");
+        sdkLanguages.put("spring", "io.swagger.codegen.languages.SpringCodegen");
+        sdkLanguages.put("csharpDotNet2", "io.swagger.codegen.languages.CsharpDotNet2ClientCodegen");
+        sdkLanguages.put("haskell", "io.swagger.codegen.languages.HaskellServantCodegen");
+    }
+
+
+    /**
+     * Generates the swagger SDK of given language
+     *
```
Comment (harsha89) [inline omod/src/main/java/org/openmrs/module/fhir/swagger/codegen/SwaggerCodeGenerator.java]: parameter descriptions missing

File: omod/src/main/java/org/openmrs/module/fhir/swagger/codegen/SwaggerCodeGenerator.java [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+        String formattedSwaggerDef=Json.pretty(swaggerDoc);
+        Path tempSdkGenDir = null;
+        File swaggerDefFile = null;
+
+        try {
+            //Create a temporary directory to store the API files
+            tempSdkGenDir = Files.createTempDirectory(ZIP_ARCHIVE_NAME_PREFIX + "_" + language + "_");
+
+            //Create a temporary file to store the swagger definition
+            swaggerDefFile =Files.createTempFile(tempSdkGenDir,
```
Comment (harsha89) [inline omod/src/main/java/org/openmrs/module/fhir/swagger/codegen/SwaggerCodeGenerator.java]: formatting issue

File: omod/src/main/java/org/openmrs/module/fhir/swagger/utils/Utils.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
+package org.openmrs.module.fhir.swagger.utils;
```
Comment (harsha89) [inline omod/src/main/java/org/openmrs/module/fhir/swagger/utils/Utils.java]: licence header is missing

File: omod/src/main/java/org/openmrs/module/fhir/swagger/codegen/SwaggerCodeGenerator.java [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+        codegenConfigurator.setInputSpec(swaggerDefLocation);
+        codegenConfigurator.setOutputDir(tempOutputDirectory);
+        codegenConfigurator.setLang(sdkLanguages.get(language));
+        final ClientOptInput clientOptInput = codegenConfigurator.toClientOptInput();
+        new DefaultGenerator().opts(clientOptInput).generate();
+    }
+
+    /**
+     * This method used to delete the directory that created during the sdk generation
+     *
```
Comment (harsha89) [inline omod/src/main/java/org/openmrs/module/fhir/swagger/codegen/SwaggerCodeGenerator.java]: parameter descriptions missing

File: omod/src/main/java/org/openmrs/module/fhir/swagger/utils/Utils.java [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+            }
+        }
+    }
+
+    private static void writeArchiveFile(File directoryToZip, List<File> fileList, String archiveLocation,
+                                         String archiveName) throws IOException {
+
+        try (FileOutputStream fileOutputStream = new FileOutputStream(archiveLocation + File.separator + archiveName
+                + ".zip");
+             ZipOutputStream zipOutputStream =new ZipOutputStream(fileOutputStream)) {
```
Comment (harsha89) [inline omod/src/main/java/org/openmrs/module/fhir/swagger/utils/Utils.java]: formatting issue space needed after =

File: omod/src/main/java/org/openmrs/module/fhir/swagger/codegen/SwaggerCodeGenerator.java [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+     * @throws IOException if something went wrong
+     */
+    public static void recursiveDeleteOnExit(Path path)throws IOException {
```
Comment (harsha89) [inline omod/src/main/java/org/openmrs/module/fhir/swagger/codegen/SwaggerCodeGenerator.java]: formatting issue

File: omod/src/main/java/org/openmrs/module/fhir/swagger/codegen/SwaggerCodeGenerator.java [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+        String tempZipFilePath = "";
+        if (swaggerDefFile.exists()) {
+
+            try (Writer swaggerFileWriter = new BufferedWriter(
+                    new OutputStreamWriter(
+                            new FileOutputStream(swaggerDefFile.getAbsoluteFile()), "UTF-8"))) {
+
+                swaggerFileWriter.write(formattedSwaggerDef);
+
+                log.debug("Writing the swagger definition was sucessful to file" + swaggerDefFile.getAbsolutePath());
```
Comment (harsha89) [inline omod/src/main/java/org/openmrs/module/fhir/swagger/codegen/SwaggerCodeGenerator.java]: space needed after file

**Comment by harsha89** [comment]
overall please add more debug logs where it necessary

**Comment by eunice18** [comment]
hello @harsha89 I will address the comments. Currently i'm working on providing support for all the languages with improvements. I will finish this soon.

**Comment by eunice18** [comment]
hello @harsha89 I addressed the comments. I will add more languages after testing.

**Comment by harsha89** [comment]
Ok @eunice18 please include all language support in your next PRs.

**Comment by kaweesi** [comment]
this breaks down the reference apps, see: https://issues.openmrs.org/browse/FM-251

File: pom.xml [hunk 11]
```diff
@@ -241,6 +242,12 @@
                 <artifactId>jackson-databind</artifactId>
                 <version>${jacksonDataBindVersion}</version>
             </dependency>
+
+            <dependency>
+                <groupId>io.swagger</groupId>
+                <artifactId>swagger-codegen</artifactId>
+                <version>${swaggerCondeGenVersion}</version>
```
Comment (kaweesi) [inline pom.xml]: this needed to exclude javax.validation, else it was resulting into https://gist.github.com/kaweesi/82c26267cc799ce8cf183f1475099ec7 which resulted into a great bug at https://issues.openmrs.org/browse/FM-246
Fixed at: https://github.com/openmrs/openmrs-module-fhir/commit/7aad32bdebdde8da3400d35d09172dbb4d4ca556

**Pull request merged**